agent-common/src/com/thoughtworks/go/agent/launcher/ServerBinaryDownloader.java

```diff
@@ -17,161 +17,134 @@
+                    Thread.sleep(period);
+                } catch (InterruptedException ie) { /* we don't care. Stupid checked exception.*/ }
             }
-            return downloaded;
         }
+        return downloaded;
     }
 
-    Map<String, String> headers() throws Exception {
-        Map<String, String> headers = serverCall.invoke(new HttpHead(checkUrl()), rootCertFile, sslVerificationMode).headers;
-        if (!headers.containsKey(MD5_HEADER)) {
-            LOG.error(format("Contacted server at URL %s but the server did not send back a response containing the header %s", downloadableFile.url(urlGenerator), MD5_HEADER));
+    void fetchUpdateCheckHeaders(DownloadableFile downloadableFile) throws Exception {
+        String url = downloadableFile.validatedUrl(urlGenerator);
+        final HttpRequestBase request = new HttpHead(url);
+        request.setConfig(RequestConfig.custom().setConnectTimeout(HTTP_TIMEOUT_IN_MILLISECONDS).build());
+
+        try (
+                CloseableHttpClient httpClient = httpClientBuilder.build();
+                CloseableHttpResponse response = httpClient.execute(request)
+        ) {
+            handleInvalidResponse(response, url);
+            this.md5 = response.getFirstHeader(MD5_HEADER).getValue();
+            this.sslPort = response.getFirstHeader(SSL_PORT_HEADER).getValue();
         }
-        return headers;
     }
 
-    private String checkUrl() {
+    protected boolean download(final DownloadableFile downloadableFile) throws Exception {
+        File toDownload = downloadableFile.getLocalFile();
+        LOG.info("download of " + toDownload + " started at " + new Date());
         String url = downloadableFile.url(urlGenerator);
-        try {
-            new URL(url);
-        } catch (MalformedURLException mue) {
-            throw new RuntimeException(
-                    "URL you provided to access Go Server: " + downloadableFile.url(urlGenerator) + " is not valid");
-        }
-        return url;
-    }
+        final HttpRequestBase request = new HttpGet(url);
+        request.setConfig(RequestConfig.custom().setConnectTimeout(HTTP_TIMEOUT_IN_MILLISECONDS).build());
 
-    private static boolean checksOut(File file, String expectedSignature) {
-        try (FileInputStream input = new FileInputStream(file)) {
-            FileDigester fileDigester = new FileDigester(input, new NullOutputStream());
-            fileDigester.copy();
-            return expectedSignature.equals(fileDigester.md5());
-        } catch (Exception e) {
-            throw new RuntimeException(e);
+        try (CloseableHttpClient httpClient = httpClientBuilder.build();
+             CloseableHttpResponse response = httpClient.execute(request)) {
+            LOG.info("got server response at " + new Date());
+            handleInvalidResponse(response, url);
+            response.getEntity().writeTo(new FileOutputStream(downloadableFile.getLocalFile()));
```

Comment:
Please check if `response.getEntity()` is non-null, and probably log it in case it is.